### PR TITLE
Add genomic data collator for llama3 with support for handling mixed case & degenerate nucleotide / control characters

### PR DESCRIPTION
**Pull request by savitha-eng**
# Add Genomic Data Collator for Mixed Case, Degenerate Base/Control CharMasking (BSHD Format)

## Summary

Adds genomic-specific data preprocessing to `llama3_native_te` recipe for training on OpenGenome2 metagenomics dataset. Implements degenerate base masking (N, R, Y, @, #) adapted from Evo2's approach, with optional uppercase label conversion for mixed-case data.

## Core Implementation

- **Add `genomic_masking_functions.py`**: Core functions from NeMo's Evo2 implementation
  - `make_upper_case()`: Vectorized ASCII-based uppercase conversion
  - `Evo2MaskingConstants`: Standard DNA tokens and control character definitions
  
- **Add `data_collator.py`**: `GenomicDataCollatorForCLM` for BSHD format
  - Masks degenerate bases (N, R, Y, etc.) by setting labels to -100
  - Masks control characters (@, #) used in data formatting
  - Optional uppercase labels (for mixed-case input → uppercase output translation)
  - Inherits from HuggingFace's `DataCollatorForLanguageModeling`
  
- **Modify `dataset.py`**: Conditional collator selection in `create_bshd_dataloader()`
  - Uses `GenomicDataCollatorForCLM` when `mask_degenerate_bases=True`
  - Uses standard collator when masking disabled (backward compatible)
  - Logs which collator is being used

## Testing

- **Add 6 collator unit tests** (`test_genomic_collator.py`):
  - `test_make_upper_case_mixed`: Verifies ASCII arithmetic uppercasing
  - `test_collator_basic`: Baseline test without masking
  - `test_collator_uppercases`: Verifies uppercase with exact value checks
  - `test_collator_masks_degenerate`: Verifies N masking with exact values
  - `test_collator_combined`: Tests both features together
  - `test_collator_handles_lowercase_degenerate`: Tests potential edge case (lowercase 'n')
  
- **Add 1 integration test** (`test_dataset.py`):
  - `test_dataloader_with_genomic_masking`: End-to-end pipeline test

## Features (Milestone 1: Metagenomics)

Based on EDA of OpenGenome2 metagenomics subset:
- **Data characteristics**: 100% uppercase, 0.001% degenerate bases (3,086 / 247M chars)
- **`mask_degenerate_bases`**: Default `True` - masks N, R, Y, @, # (minimal impact)
- **`uppercase_labels`**: Default `False` - not needed for metagenomics (already uppercase)
- **Backward compatible**: Existing code works unchanged (flags default to standard collator)

## Next Steps (Milestone 2: Full Dataset)

- Add phylogenetic tag masking (`|d__Bacteria;p__Proteobacteria|`) for full OpenGenome2
- Enable `uppercase_labels` for mixed-case datasets
- Add THD (sequence packing) format for memory efficiency

## Implementation Details

**Masking approach** (HuggingFace vs Evo2):
- **Evo2/Megatron**: Uses separate `loss_mask` tensor (0/1 values)
- **Ours/HuggingFace**: Sets `labels=-100` (PyTorch's built-in `ignore_index`)
- Both achieve the same result via different mechanisms

### Description

This PR adds genomic-specific data preprocessing for training Llama3 models on OpenGenome2 metagenomics data. The collator masks degenerate bases and control characters that shouldn't be predicted during training, following the approach used in Evo2.

The implementation is minimal, backward compatible, and well-tested. It only affects training when explicitly enabled via config flags.

#### Usage

Enable genomic masking in your training config:

```yaml
# In hydra config (e.g., train_8b_metagenome.yaml)
dataset:
  mask_degenerate_bases: true  # Mask N, R, Y, @, #
  uppercase_labels: false      # Not needed for metagenomics (already uppercase)
```

Or pass programmatically:

```python
from dataset import create_bshd_dataloader

dataloader, _ = create_bshd_dataloader(
    distributed_config=config,
    tokenizer_path="./example_checkpoint",
    load_dataset_kwargs={...},
    micro_batch_size=2,
    mask_degenerate_bases=True,  # Enable degenerate masking
    uppercase_labels=False,      # Disable for metagenomics
)
```

The collator automatically handles masking in the data pipeline - no model changes needed.

---

### Type of changes

- [x] New feature (non-breaking change which adds functionality)

---

### Pre-submit Checklist

- [x] I have tested these changes locally (17/17 tests passing)
- [x] I have updated the documentation accordingly (docstrings in code)
- [x] I have added/updated tests as needed (6 unit + 1 integration test)
- [x] All existing tests pass successfully (backward compatible)

### CODE REVIEW

File: bionemo-recipes/recipes/llama3_native_te/data_collator.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import torch
+from genomic_masking_functions import make_upper_case
+from transformers import PreTrainedTokenizerBase
+from transformers.data.data_collator import DataCollatorForLanguageModeling
+
+
+logger = logging.getLogger(__name__)
+
+
+class GenomicDataCollatorForCLM(DataCollatorForLanguageModeling):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/data_collator.py]: it's going to be easier to extend this for THD if you restructure this a bit

```python
@dataclass
class GenomicDataCollator
    base_collator: DataCollator
    dna_tokens: list[int] = [65, 67, 71, 84, 97, 99, 103, 116]
    control_tags: list[int] = [64, 35] # '@', '#'

    def __call__(self, features: list) -> dict[str, Any]:
	    # Captive collator handles batching and CLM label creation
        batch = self.base_collator(features)

        labels = batch["labels"]

        # Step 1: Uppercase labels (inputs stay mixed case)
        if self.uppercase_labels:
            labels, _ = make_upper_case(labels)

        # Step 2: Mask degenerate bases and control characters
        if self.mask_degenerate_bases:
            dna_tokens_tensor = torch.tensor(self.dna_tokens, device=labels.device)
            control_tensor = torch.tensor(self.control_tags, device=labels.device)

            # Identify non-DNA tokens
            not_dna = ~torch.isin(labels, dna_tokens_tensor)
            is_control = torch.isin(labels, control_tensor)

            # Mask both, but preserve existing -100 values
            labels[(not_dna | is_control) & (labels != -100)] = -100

        batch["labels"] = labels
        return batch
```
